### PR TITLE
Added FWSE format.

### PR DESCRIPTION
**Pull request by PredatorCZ**
Can be now used as `add_subdirectory` for CMake.

### CODE REVIEW

**Comment by CyberBotX** [comment]
Other than the change in the root CMakeLists.txt, I'm not sure I see why the other CMake changes are necessary? The existing paths should be working fine, I've seen no issues with them as they are.

**Comment by bnnm** [comment]
Code looks fine (can't comment on CMake stuff) but if you could try to merge first, I'm no good at that.
Also maybe rename `cpm_ima_expand_nibble` to `mft_ima_expand_nibble` for consistency.

**Comment by PredatorCZ** [comment]
> 
> 
> Other than the change in the root CMakeLists.txt, I'm not sure I see why the other CMake changes are necessary? The existing paths should be working fine, I've seen no issues with them as they are.

They use CMAKE_SOURCE_DIR, this won't work, if they're used as subdirectory. Since this variable will be always location of very root CMakeLists.txt.
I can modify `set(VGMSTREAM_EXT_INCLUDES ${CMAKE_CURRENT_SOURCE_DIR}../ext_includes)` for better clarification or replace this var directly with `${CMAKE_CURRENT_SOURCE_DIR}../ext_includes/...`, whichever you prefer better.

**Comment by CyberBotX** [comment]
> > Other than the change in the root CMakeLists.txt, I'm not sure I see why the other CMake changes are necessary? The existing paths should be working fine, I've seen no issues with them as they are.
> 
> They use CMAKE_SOURCE_DIR, this won't work, if they're used as subdirectory. Since this variable will be always location of very root CMakeLists.txt.
> I can modify `set(VGMSTREAM_EXT_INCLUDES ${CMAKE_CURRENT_SOURCE_DIR}../ext_includes)` for better clarification or replace this var directly with `${CMAKE_CURRENT_SOURCE_DIR}../ext_includes/...`, whichever you prefer better.

I understand that CMAKE_SOURCE_DIR is the directory of the root CMakeLists.txt, but I was referencing ext_includes relative to that directory, so that's why I'm confused as to how this won't work. It's been working fine for me from what I've seen. In what cases does it not work out?

**Comment by PredatorCZ** [comment]
> I understand that CMAKE_SOURCE_DIR is the directory of the root CMakeLists.txt, but I was referencing ext_includes relative to that directory, so that's why I'm confused as to how this won't work. It's been working fine for me from what I've seen. In what cases does it not work out?

When I use it as `add_subdirectory(modules/vgmstream)`.

**Comment by CyberBotX** [comment]
> > I understand that CMAKE_SOURCE_DIR is the directory of the root CMakeLists.txt, but I was referencing ext_includes relative to that directory, so that's why I'm confused as to how this won't work. It's been working fine for me from what I've seen. In what cases does it not work out?
> 
> When I use it as `add_subdirectory(modules/vgmstream)`.
> Or I'm doing something wrong.

Where are you doing the add_subdirectory from and what problems are being caused by the current CMake code? I might be able to help figure out why but not without knowing what the problem actually is, because it doesn't make sense to me currently.

**Comment by PredatorCZ** [comment]
Please check https://github.com/PredatorCZ/RevilToolset/blob/0f8a1bb4061d43775e6cd3cca16f94b462e7880a/SPACConvert/CMakeLists.txt#L35

This is how I use it.

**Comment by CyberBotX** [comment]
Ah, I see now. OK, then yes, please make sure all paths use `CMAKE_CURRENT_SOURCE_DIR` to keep things relative. Not using that could lead to some subtle errors I believe. If you are going to do this, please do it throughout the project. Also, I believe you `CMAKE_CURRENT_SOURCE_DIR` does not include a trailing slash so be sure to put a slash after it.

**Comment by PredatorCZ** [comment]
Added more relativity.
Tested with WIN32 and install command.
Following are untested:
Maiattrac3+
Foobar plugins (uses vcxproj)
Audacitous

Also a question, where do you want to generate mp4v2 config header?
I believe that you use CMAKE_SOURCE_DIR == CMAKE_BINARY_DIR, not everyone uses such approach, so you want to keep it in source dir, or move it to the binary dir?

**Comment by CyberBotX** [comment]
The reason that I made the mp4v2 header go to the binary dir is because I focused on out-of-source builds to keep the source tree pristine. Writing to the source directory is a bad approach in my opinion, as it makes source control harder to deal with.

**Comment by PredatorCZ** [comment]
Yes, I totally agree. I just got a bit confused by CI builds and the fact, that you generate version files in the source dir. Changed to binary dir.

File: cli/CMakeLists.txt [hunk 1]
```diff
@@ -17,8 +17,8 @@ if(WIN32)
 
 	# Make sure that the binary directory is included (for version.h), as well as the getopt library include directory
 	target_include_directories(vgmstream_cli PRIVATE
-		${CMAKE_BINARY_DIR}
-		${CMAKE_SOURCE_DIR}/ext_libs/Getopt)
+		${VGM_SOURCE_DIR}
```
Comment (CyberBotX) [inline cli/CMakeLists.txt]:
```suggestion
		${VGM_BINARY_DIR}
```

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -41,7 +41,8 @@ setup_target(libvgmstream)
 
 # Set up the proper include directories
 target_include_directories(libvgmstream PRIVATE
-	${CMAKE_CURRENT_SOURCE_DIR}
+	${VGM_SOURCE_DIR}
```
Comment (CyberBotX) [inline src/CMakeLists.txt]:
```suggestion
	${VGM_BINARY_DIR}
```

File: winamp/CMakeLists.txt [hunk 3]
```diff
@@ -29,7 +29,7 @@ target_compile_definitions(in_vgmstream PRIVATE
 	_UNICODE)
 
 # Make sure that the binary directory is included (for version.h)
-target_include_directories(in_vgmstream PRIVATE ${CMAKE_BINARY_DIR})
+target_include_directories(in_vgmstream PRIVATE ${VGM_SOURCE_DIR})
```
Comment (CyberBotX) [inline winamp/CMakeLists.txt]:
```suggestion
target_include_directories(in_vgmstream PRIVATE ${VGM_BINARY_DIR})
```

File: src/CMakeLists.txt (hunk 2 above)
Comment (PredatorCZ) [reply src/CMakeLists.txt]: Doesn't seem that this lib requires such include.
I can build it without this path, please confirm.

Comment (CyberBotX) [reply src/CMakeLists.txt]: If it build fine without it, then it is fine to leave it out. I could've sworn there was a reason I had it set... (and it should've been SOURCE, not BINARY, my suggested change is wrong in this case).

Comment (PredatorCZ) [reply src/CMakeLists.txt]: It's already removed.
I think this MR can be resolved.

**Pull request merged**